programs/stake/src/stake_state.rs

```diff
@@ -705,6 +784,194 @@ pub fn redelegate(
     Ok(())
 }
 
+pub fn move_stake(
+    invoke_context: &InvokeContext,
+    transaction_context: &TransactionContext,
+    instruction_context: &InstructionContext,
+    source_account_index: IndexOfAccount,
+    lamports: u64,
+    destination_account_index: IndexOfAccount,
+    clock: &Clock,
+    stake_history: &StakeHistory,
+    stake_authority_index: IndexOfAccount,
+) -> Result<(), InstructionError> {
+    let mut source_account = instruction_context
+        .try_borrow_instruction_account(transaction_context, source_account_index)?;
+
+    let mut destination_account = instruction_context
+        .try_borrow_instruction_account(transaction_context, destination_account_index)?;
+
+    let (source_merge_kind, destination_merge_kind) = move_stake_or_lamports_shared_checks(
+        invoke_context,
+        transaction_context,
+        instruction_context,
+        &source_account,
+        lamports,
+        &destination_account,
+        clock,
+        stake_history,
+        stake_authority_index,
+    )?;
+
+    // source and destination will have their data reassigned
```

Comment:
I don't really understand this comment in this location

Reply:
hmm i can reword or just delete it, we check the buffer lengths here because we need to write new `StakeState` objects because `Delegation`s change. so if/when there is a new `StakeStateV3` thats longer, it will fail for old accounts that havent been reallocated from their `StakeStateV2` length (tho i guess a new stake account format would come with a batch process in the bank that reallocs all of them)

Reply:
I see, so you're distinguishing the reqs for reserializing here from deserialization above (in the `get_state()` calls). I do think it's unlikely that we'll want to the let the type or multiple types we deserialize deviate from what we write, but I suppose these checks do no harm. Maybe just make the comment a little more explicit, then? Something like:
```suggestion
    // Ensure source and destination data are the right size for the StakeState type to be written.
    // This is a safeguard against potential future changes to support deserializing multiple
    // StakeState types of different lengths.
```
(Sorry, not wrapped to the right num chars.)